rc/tools/lint.kak

```diff
@@ -216,7 +216,16 @@ define-command \
         cat "$dir"/result > "$dir"/fifo
         rm -rf "$dir"
 
-        } & ) >"$dir"/stderr 2>&1 </dev/null
+        # We don't need the cleanup hook any more, the files are gone
+        printf "eval -buffer '%s' 'remove-hooks buffer lint-cleanup'" "$kak_bufname" | kak -p "$kak_session"
+
+        } & ) >"$dir"/stderr 2>&1 </dev/null &
+
+        readonly pid_async_worker=$!
+        # Make sure temporary files are cleaned up, asynchronous processes killed
+        printf "hook -once -always -group lint-cleanup buffer BufClose .* '
+            nop %%sh{ rm -rf \"%s\"; kill %d }
+        '" "$dir" "$pid_async_worker"
```

Comment:
This looks dangerous, is it guaranteed that $pid_async_worker still points to that process here and has not been reused ?

Reply:
I think somebody more knowledgeable than me about how shell works should confirm that.

[POSIX specification](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_05_02):

> !
> Expands to the decimal process ID of the most recent background command (see Lists) executed from the current shell. (For example, background commands executed from subshells do not affect the value of "$!" in the current shell environment.) For a pipeline, the process ID is that of the last command in the pipeline.

I assume `( … ) & pid=$!; …; kill $pid` should behave as expected.

Reply:
In general, there's no way to guarantee that a PID obtained at time X still refers to the same process at time Y. If you use the Linux-specific syscall [`pidfd_open()`](https://man7.org/linux/man-pages/man2/pidfd_open.2.html) to obtain a permanent process identifier, then a later call to `pidfd_send_signal()` will fail if the process has already exited, even if the PID has been re-used. However, there's still a chance that the PID has been recycled between the time it is obtained and the time `pidfd_open()` is called

Reply:
The PIDs on my system get into the nine digits over time, so I assumed they could not be re-used.

It's also a common pattern in service management to save the PID of a server into a `.pid` file, and try `kill`ing that number at an arbitrary, later time. Is that pattern wrong?

Are reusable PIDs really common in the wild, or is that irrelevant because we might end up killing the wrong process due to how linux works anyway? What about other kernels?